requirements.txt

```diff
@@ -1,11 +1,7 @@
-docker==4.1.0
-pyyaml==5.1.2
-requests==2.22.0
+docker==4.2.0
```

Comment:
there are two ways to do this, pin requirements to exact version (like we do now), or use < %next major version%.
first approach is more stable, but we have to do this kind of PR every once in a while
second approach does not require this (or only for major versions), but breaks easier, because we sometimes rely on internals, which can break even in minor version updates
how other great and cool python projects handle this?
I am leaning towards first approach, what are your thoughts?

Reply:
I'm not aware about concrete Python best practices. When dealing with quite non-stable libraries environment I personally like npm approach with `package.json.lock` file. It's the first approach you mentioned but with tooling automation. Here we simulate it by hand. In more stable environments like Java I think it's OK to have version ranges (in my previous projects we usually did).